Add an API Client package

zcahana opened the pull request:
A client package aimed to be used by other components (controller, sidecar, tests) that interact with the registry over HTTP.

- This package currently depends on the 'api/protocol/amalgam8' package which defines the API endpoint paths.
- This package redeclares several types defined in the 'api/protocol/amalgam8' package (ServiceInstance, etc), so that a consumer of the client can interact with a single package only.

elevran commented on client/client.go:
```diff
@@ -0,0 +1,231 @@
+)
+
+// Client defines the interface used by clients of the Amalgam8 Service Registry.
+type Client interface {
```
client.Client is redundant. Should reconsider name of package (e.g., registry.Client, or am8sd.Client)

elevran commented on client/client.go:
```diff
@@ -0,0 +1,231 @@
+	Register(instance *ServiceInstance) (*ServiceInstance, error)
+	Deregister(id string) error
+	Renew(id string) error
+	ListServices() ([]string, error)
```
Consider if it makes sense to split into two separate interfaces, one for producer (register, renew, delete) and one for consumer (list...)

elevran commented on client/client.go:
```diff
@@ -0,0 +1,231 @@
+	Deregister(id string) error
+	Renew(id string) error
+	ListServices() ([]string, error)
+	ListInstances(instanceFilter *InstanceFilter) ([]*ServiceInstance, error)
```
Do you need a *InstanceFilter? name of parameter can be 'filter'.

elevran commented on client/client.go:
```diff
@@ -0,0 +1,231 @@
+	"time"
+)
+
+// Client defines the interface used by clients of the Amalgam8 Service Registry.
```
Should probably document the different methods as well, at least at a high level. It is not immediately clear why three list* methods are needed.

elevran commented on client/client.go:
```diff
@@ -0,0 +1,231 @@
+	httpClient *http.Client
+}
+
+func NewRESTClient(config ClientConfig) (*RESTClient, error) {
```
NewClient would suffice. RESTClient is, IMHO, an implementation detail.

elevran commented on client/filter.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+// InstanceFilter is used to filter service instances returned from lookup calls.
+type InstanceFilter struct {
```
Consider pulling all exposed interfaces and types into an api.go file. RESTClient would go into a separate file.

elevran commented on client/filter.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+// asQueryParams convert the filter into a set of query parameters that can be added to a lookup request.
+func (filter *InstanceFilter) asQueryParams() url.Values {
```
struct would do as well

elevran commented on client/instance.go:
```diff
@@ -0,0 +1,46 @@
+//
+// It is used both as an input to registration calls, as well as the output of discovery calls.
+// Depending on the context, some of the fields may be mandatory, optional, or ignored.
+type ServiceInstance struct {
```
See comment on filter.go

zcahana commented:
@elevran I've added 2 commits addressing your comments.
I haven't found a much better name for the `Client` interface, but I did split it into a `Registry` and `Discovery` interfaces (unified by the `Client` interface), so we can expect consumers of this package to mostly refer to `client.Registry` or `client.Discovery`, and not `client.Client`.

elevran commented:
lgtm, will merge.
